smithy-model/src/main/java/software/amazon/smithy/model/transform/ScrubTraitDefinitions.java

```diff
@@ -42,14 +45,21 @@
  */
 final class ScrubTraitDefinitions {
     Model transform(ModelTransformer transformer, Model model) {
+        return transform(transformer, model, FunctionalUtils.alwaysTrue());
+    }
+
+    Model transform(ModelTransformer transformer, Model model, Predicate<Shape> keepFilter) {
         // Find all trait definition shapes and private shapes in the prelude.
-        Set<Shape> toMark = Stream.concat(
+        Set<Shape> traitDefinitions = Stream.concat(
                 model.shapes().filter(shape -> shape.hasTrait(TraitDefinition.class)),
                 model.shapes().filter(shape -> Prelude.isPreludeShape(shape) && shape.hasTrait(PrivateTrait.class))
         ).collect(Collectors.toSet());
 
+        // Filter any trait definitions from being marked for removal that we want to retain.
+        Set<Shape> toMark = traitDefinitions.stream().filter(keepFilter).collect(Collectors.toSet());
```

Comment:
Can this filtering be done all on a single stream instead of needing to make two sets?